plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+    assertRedisErr(redisAsyncSetDisconnectCallback(context, disconnectCB), "failed to set disconnect callback");
+    selectDb(ctx);
+}
+void AsyncConnection::assertConnection()
+{
+    if (!context)
+        rtlFail(0, "Redis Plugin: async context mem alloc fail.");
+    assertContextErr(context);
+}
+void AsyncConnection::assertContextErr(const redisAsyncContext * context)
+{
+    if (context && context->err)
+    {
+        const AsyncConnection * connection = (const AsyncConnection*)context->data;
+        if (connection)
+        {
+            VStringBuffer msg("Redis Plugin : failed to create connection context for %s:%d - %s", connection->getMaster(), connection->getPort(), context->errstr);
+            rtlFail(0, msg.str());
+        }
+        {
+            VStringBuffer msg("Redis Plugin : failed to create connection context - %s", context->errstr);
+            rtlFail(0, msg.str());
+        }
+    }
+}
+void AsyncConnection::connectCB(const redisAsyncContext * context, int status)
+{
+    if (status != REDIS_OK)
+    {
+        if (context->data)
+        {
+            const AsyncConnection * connection = (AsyncConnection *)context->data;
+            VStringBuffer msg("Redis Plugin : failed to connect to %s:%d - %s", connection->master.str(), connection->port, context->errstr);
+            rtlFail(0, msg.str());
+        }
+        else
+        {
+            VStringBuffer msg("Redis Plugin : failed to connect - %s", context->errstr);
+            rtlFail(0, msg.str());
+        }
+    }
+}
+void AsyncConnection::disconnectCB(const redisAsyncContext * context, int status)
+{
+    if (status != REDIS_OK)//&& context->err != 2)//err = 2:  ERR only (P)SUBSCRIBE / (P)UNSUBSCRIBE / QUIT allowed in this contex
+    {
+        if (context->data)
+        {
+            const AsyncConnection  * connection = (AsyncConnection*)context->data;
+            VStringBuffer msg("Redis Plugin : server (%s:%d) forced disconnect - %s", connection->master.str(), connection->port, context->errstr);
+            rtlFail(0, msg.str());
+        }
+        else
+        {
+            VStringBuffer msg("Redis Plugin : server forced disconnect - %s", context->errstr);
+            rtlFail(0, msg.str());
+        }
+    }
+}
+AsyncConnection * AsyncConnection::createConnection(ICodeContext * ctx, const char * options, unsigned __int64 database)
+{
+    return new AsyncConnection(ctx, options, database);
+}
+void AsyncConnection::assertRedisErr(int reply, const char * _msg)
+{
+    if (reply != REDIS_OK)
+    {
+        VStringBuffer msg("Redis Plugin: %s", _msg);
+        rtlFail(0, msg.str());
+    }
+}
+void AsyncConnection::assertOnError(const redisReply * reply, const char * _msg)
+{
+    if (!reply)
+    {
+        assertConnection();
+        //There should always be a connection error
+        VStringBuffer msg("Redis Plugin: %s%s", _msg, "no 'reply' nor connection error");
+        rtlFail(0, msg.str());
```

Comment:
Should be using an error code other than 0 for all new errors that we report (and going back and adding sensible codes for all previous ones)